eva/parser/lark_visitor/_table_sources.py

```diff
@@ -0,0 +1,274 @@
+# coding=utf-8
+# Copyright 2018-2022 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from lark import Token, Tree
+
+from eva.expression.tuple_value_expression import TupleValueExpression
+from eva.parser.select_statement import SelectStatement
+from eva.parser.table_ref import Alias, JoinNode, TableRef, TableValuedExpression
+from eva.parser.types import JoinType
+from eva.utils.logging_manager import logger
+
+##################################################################
+# TABLE SOURCES
+##################################################################
+
+
+class TableSources:
+    def select_elements(self, tree):
+        kind = tree.children[0]
+        if kind == "*":
+            select_list = [TupleValueExpression(col_name="*")]
+        else:
+            select_list = []
+            for child in tree.children:
+                element = self.visit(child)
+                select_list.append(element)
+        return select_list
+
+    def table_sources(self, tree):
+        return self.visit(tree.children[0])
+
+    def table_source(self, tree):
+        left_node = None
+        join_nodes = []
+
+        for child in tree.children:
+            if isinstance(child, Tree):
+                if child.data == "table_source_item_with_sample":
+                    left_node = self.visit(child)
+                    join_nodes = [left_node]
+                elif child.data.endswith("join"):
+                    table = self.visit(child)
+                    join_nodes.append(table)
+
+        num_table_joins = len(join_nodes)
+
+        # Join Nodes
+        if num_table_joins > 1:
+            # Add Join nodes -> left deep tree
+            # t1, t2, t3 -> j2 ( j1 ( t1, t2 ), t3 )
+            for i in range(num_table_joins - 1):
+                join_nodes[i + 1].join_node.left = join_nodes[i]
+
+            return join_nodes[-1]
+        else:
+            return join_nodes[0]
+
+    def table_source_item_with_sample(self, tree):
+        sample_freq = None
+        alias = None
+        table = None
+
+        for child in tree.children:
+            if isinstance(child, Tree):
+                if child.data == "table_source_item":
+                    table = self.visit(child)
+                elif child.data == "sample_clause":
```

Comment:
`Antlr` supported something like https://github.com/georgia-tech-db/eva/blob/dc165eb2fb6cbe330340b53c3d7b099e809bb79c/eva/parser/parser_visitor/_table_sources.py#L93
Just curious if lark also has something similar. `antlr` approach looked cleaner to me.
Btw, how do we know the possible values of `data`? Do we get it from the autogenerated file (if any) or need to look at the grammer?

Reply:
Lark unfortunately does not support `alias` for `rule particles`. It only supports `alias` for full `rule`s.

Reply:
Lark does not support aliases within a rule -- it only supports an alias for the entire rule.

I looked at the grammar file to write the transformation function.